Split source / output

WolfieWerewolf opened the pull request:
Hi guys.. I am very impressed with the work you have done here; I wish I had known about it sooner.  Once I got up and running (thanks to @zicklag) I thought I might be able to help out by contributing a few small changes.
- Moved the hard coded webidl.Options to webidl.json in the project folder.
- Added an optional field to webidl.Options for 'out' so we can specify the output folder.
- Added an OUT field to the Makefile (ideall we want this all in one place but since we are using make I thought I would just follow the existing approach)
- Modified Generate.hx to use the opts.out field to put the generated output into a target folder.. in my example it's ./webRoot.
- Added a make clean to remove the generated output from webRoot

It should be possible to simply execute "make js" and have everything dump into webRoot.  If you run something like live-server or the python simple server in there you should get the running wasm.

I have tested this on Linux Mint which seems entirely compatible with Ubuntu based on the tested / help that @zicklag provided.  I don't have access to OSX / MacOS or Windows unfortunately.

If you are happy with this great, if there are changes you would like, I am happy to make them, if you don't like any of this then that's ok too ;-) I just wanted to see if I could help out.
Thanks for your consideration.
Kind Regards
Wolfie

zicklag commented:
I haven't tested running it, but I think its a good idea. :+1:

My only thought from looking at it is that we could maybe merge the splitout sample with the other one, or at least put them both in a samples folder.

WolfieWerewolf commented:
@zicklag Excellent.. I would be more than happy to merge it into the existing sample.. I just didn't want to go ahead and do that unless I had some kind of indication from you guys that it would be acceptable to do so.  I'll manually make the changes and let you know when that's done.. thanks.

zicklag commented:
Also you might want to wait to hear from @ncannasse. Its his repo, I've just contributed. :smiley:

WolfieWerewolf commented:
@zicklag ... cool, @ncannasse with your permission I would like to do as @zicklag suggested :-)

WolfieWerewolf commented:
@zicklag ... while I have you here maybe you can answer something for me?  What is the plan with regards to handling bindings for multiple c++ classes in the same project.  The idl file in the sample is called point.idl presumably because the only thing it handles presently is the Point class but if I wanted to  add a class (different header and implementation file) would I just append it to the same .idl file?  And if so I  suspect I will have to handle it somehow in the makefile?  Perhaps make is a bit light for a larger workload if that's the case? Perhaps the work presently being done by make should be handled by a webidl executable?  I'm not sure what the long term goals are for this project but it looks like something that I would like to make more industrial use of than just a few samples...

WolfieWerewolf commented:
@zicklag well that was easier than I expected.  I've gone ahead and updated my fork with the changes necessary to compile multiple c++ targets and generate the bindings.  For now I updated the splitout example. 

- Handling parsing of #includes (Generalte line 92.. Haxe is refreshingly Functional these days.. )in generator so it is no longer necessary to supply them manually in the webidl.json file.  I haven't tested cases where there might be duplicate includes so I have a note to test that and deduplicate the array if that happens (I suspect it will). 
- Added a Array<String> field for sourceFiles in webidl.json and removed the hard coded source files in the sample project

From a user perspective:
- All you have to do is ensure the sourceFiles in webidl.json are supplied in the array; it is no longer necessary to call out the includes specifically, it should parse / generate them dynamically now.
- Update the Makefile supplying the cpp files on the libpoint.hdll and (libpoint_win lines if on Windows)

The makefile isn't terrible but I think it could still become troublesome having to make changes in two places.. I'll put some more thought into that.
If you guys could test that and let me know what you think I would appreciate it.
After you run make.js and launch a webserver in ./webRoot you should see the following output on the js console in your browser (tested on Chrome latest).
```js
splitout/Splitout.hx:17: Result = 11,13 len=17.029386365926403
libpoint.js:8 This is Context!
libpoint.js:8 This is test!
```

zicklag commented:
Sounds cool! I'm a little busy, but I'll let you know if I try it.

ncannasse commented:
Hey there ! I'm glad there's interest in this project. ATM I have already many (=too much) repositories to monitor so i'm not sure I can invest a lot of time on webidl. 

So if @zicklag wants to be added as a contributor and keep developing it, I will gladly give the corresponding rights :) I'm planning to use webidl for bullet HL support, as long as it doesn't break things here it's fine for me.

WolfieWerewolf commented:
@ncannasse .. Hi Nick.. very good news, congratulations @zicklag you have been promoted!.  I'll go ahead and make the changes you suggested.  Thank you both.

zicklag commented:
@ncannasse That would be great, thanks!

WolfieWerewolf commented:
Hi @zicklag .. as per our discussion I have gone ahead and merged the samples as follows:
- Created a top level samples folder in the project root
- Renamed the existing sample to simple
- The simple sample is an aggregation of the original and the changes described about in this PR.

I've additionally taken the liberty of changing the naming a little based on our discussions:
- The Module name reflects the folderName+Module.hx --> SimpleModule.hx
- The Entry point reflects the folder name as Upper Case -> Simple.hx
- The IDL file reflects the folderName since the idl file is more related to the module than it's contents.

When generating output the following naming is used:
- We prefix the folder name with lib_ (snake case) following the examples by the emscripten guys since this is the generated js.
- The custom js bindings are held in folder.js -> simple.js

I guess you are the new maintainer?  If so could you please test at your convenience and let me know if any changes are needed or not.  If not I am happy for you to merge this as it is.  I will have further enhancements to contribute later this week.

@ncannasse .. Nick, you said something important above "so long as we don't break anything".. can you be specific about what that means?  What can I test against to ensure that nothing is broken and / or if there are breaking changes what can I use to document against to keep you current?

Thanks
Kind Regards
/W

zicklag commented:
@WolfieWerewolf I'll check it out as soon as I can. :slightly_smiling_face: 

As far as what we can't break, it's [bullet bindings for HashLink](https://github.com/HaxeFoundation/hashlink/tree/master/libs/bullet).

WolfieWerewolf commented:
@zicklag ... excellent thanks.. I'll test against that shortly.  There will likely need to be some changes to those bindings to make use of this PR.. if so I'll create a related PR for that project.

WolfieWerewolf commented:
@zicklag .. mate.. I haven't tried to compile this yet but it looks to me like the Haxe Bullet bindings include a fork of the webidl that it was using at the time and doesn't reference this project?
https://github.com/armory3d/haxebullet/tree/master/Sources/webidl
And the webidl files look to be about 3 months old.  I would suspect that various changes, not just the ones in this PR, could have had an impact on it.  Has anyone tested this recently?

zicklag commented:
That's not the repo I was talking about here. The one I'm talking about is the one that HashLink itself comes with: https://github.com/HaxeFoundation/hashlink/tree/master/libs/bullet.

BTW I've got a PR for Armory3D's Haxebullet to update it. That is part of the Kha integration I mentioned in the other issue. ;) https://github.com/armory3d/haxebullet/pull/27

WolfieWerewolf commented:
oops.. my bad.   I'm sorry I'm not following there's a lot going on.  Do you still want me to verify against the bullet stuff in hashlink (using the correct URL this time)... sorry for my confusion.. I am just getting caught up with all the dependencies etc..

zicklag commented:
No problem. You could test it if you want. It should still work fine because I think it pretty much just uses the generated [bullet.cpp](https://github.com/HaxeFoundation/hashlink/blob/master/libs/bullet/bullet.cpp) file from WebIDL and our changes have been mostly to the JavaScript generation and the export locations, which aren't used by that repo directly anyway. Still could be good to test. I think the `bullet.cpp` file should actually end up identical to the one already there if you generate bindings with the included WebIDL file.

WolfieWerewolf commented:
Ok good info thanks.. I will test just to be sure.. I did make some changes to the way the configuration is handled though.. I moved the static config out of the Module to a json file so that might need some tweaking.. thanks for the info

zicklag commented on samples/simple/SimpleModule.hx:
```diff
@@ -0,0 +1,28 @@
+
+class SimpleModule {
+
+	static var json = {
```
I'm thinking we can simplify this block like so:

```suggestion
	static var json = haxe.Json.parse(sys.io.File.getContent("./webidl.json"));
```

zicklag replied:
The GitHub auto-commit thing won't remove the 4 lines below, so we'll need to do that manually.

zicklag commented on .vscode/c_cpp_properties.json:
```diff
@@ -0,0 +1,16 @@
+{
```
I'm assuming this file wasn't meant to be in here as you added `.vscode` to the `.gitignore`?

WolfieWerewolf replied:
Correct.. I should not have pushed that.

zicklag commented on samples/simple/index.html:
```diff
@@ -0,0 +1,9 @@
+<!doctype html>
```
Do we need this `index.html` and the one in `webRoot`?

WolfieWerewolf replied:
Well I like to not edit anything in webRoot directly and treat it all as generated.. You'll notice that the new build copies the index.html from the source root (which is still the project root) every time you compile.. so changes should be made in sources not in webRoot (output).

zicklag commented on samples/simple/webidl.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "idlFile": "simple.idl",
+    "nativeLib": "lib_simple",
+    "sourceFiles": ["point.cpp", "context.cpp"],
```
Maybe we should add a `sourcesDir` option instead of `sourceFiles`. That is how I did the Kha integration. It is really difficult if you have a library like Bullet and you have to list each .cpp file manually.

WolfieWerewolf replied:
Yeah that's the intent but I didn't want to do all that in one PR... The next step would be to create a sourceDir...

zicklag commented on webidl/Generate.hx:
```diff
@@ -99,10 +88,19 @@ template<typename T> pref<T> *_alloc_const( const T *value ) {
+		for(sFile in opts.sourceFiles){
 			add("");
-			add(StringTools.trim(opts.includeCode));
+			for(l in sys.io.File.getBytes(if(sys.FileSystem.exists(opts.out+ "/" + sFile)){
```
I find this block of code very difficult to read. This is a bit cleaner:


```haxe
var path = null;
for(sFile in opts.sourceFiles){
	add("");
	path = FileSystem.exists(opts.out + "/" + sFile) ? opts.out + "/" + sFile : sFile;
	for(l in sys.io.File.getContent(path).split("\n")){
		if(StringTools.startsWith(l, "#include")){
			add(StringTools.trim(l));
		}
	}
}
```

Also for large projects would this be a performance concern, going through all of the source files that could represent a large library? I don't know how long that takes, it probably isn't a problem, but I thought I'd bring it up. Maybe we should test it with a bunch of files or add a `autoIncludeIncludes` ( :wink: ) flag or something like that to the `webidl.json` ( maybe with a better name like `detectIncludes` ).

WolfieWerewolf replied:
Looks good to me.. I'm coming from Kotlin and I've been embracing it's functional side so I've become used to long functions.. however you want it man is fine with me.

zicklag commented on webidl/Generate.hx:
```diff
@@ -399,7 +399,7 @@ template<typename T> pref<T> *_alloc_const( const T *value ) {
 			if( p.substr(0, 2) == "-O" )
 				hasOpt = true;
 		if( !hasOpt )
-			params.push("-O2");
```
I think `O2` is probably a good default. I'm not really set on it if you have a good reason, but I think `O2` made sense.

WolfieWerewolf replied:
And that is also an oversight on my part (good catch).. I dropped down to -O0 so I could more easily read the output.. you are correct this should be optimised.

zicklag commented on webidl/Generate.hx:
```diff
@@ -412,26 +412,32 @@ template<typename T> pref<T> *_alloc_const( const T *value ) {
 		for( cfile in sources ) {
-			var out = cfile.substr(0, -4) + ".bc";
-			var args = params.concat(["-c", cfile, "-o", out]);
+			var sourcePath = if(sys.FileSystem.exists(opts.out+ "/" + cfile)){
```
I might be missing something, but why are we looking in the output directory as a possible location for a source file?

WolfieWerewolf replied:
Because it is, the only place where the lib_project.cpp is going to ever exist is in the output directory; it's generated.

zicklag replied:
In order to preserve the sources separate from the output dir concept, which I think is good, we could create like a "bundled" directory maybe and add that to the `webidl.json` options?

zicklag replied:
That makes sense, in that case I would just update the if statement to be single lined:

```haxe
var sourcePath = sys.FileSystem.exists(opts.out + "/" + cfile) ? opts.out + "/" + cfile : cfile;
````

zicklag commented on webidl/Generate.hx:
```diff
@@ -412,26 +412,32 @@ template<typename T> pref<T> *_alloc_const( const T *value ) {
 		sys.io.File.saveContent(tmp, output);
 		command("make", ["-f", tmp]);
 		sys.FileSystem.deleteFile(tmp);
+		sys.io.File.copy("index.html", opts.out+"/index.html");
```
I see that this explains the two `index.html` files. If we have the favicon already in the webRoot folder when it builds I don't see any reason that we can't have the `index.html` file in there as well and leave out this copy.

WolfieWerewolf replied:
Sure.. I don't mind either way..

zicklag replied:
In order to preserve the sources separate from the output dir concept, which I think is good, we could create like a "bundled" directory maybe and add that to the `webidl.json` options?

WolfieWerewolf commented:
Mate.. that was an excellent code review.. thank you..

zicklag commented:
No problem. That was actually the first time I've ever reviewed a PR to a repo that I maintain.

WolfieWerewolf commented:
Nice.. I do appreciate it because no matter how much I look at code I'll always miss something.. like the .vscode thing... I'll go through this again with our comments, make the necessary changes and push that up.. likely tomorrow mate.. thanks again.
